File: 2024/01/solution.py
from collections import Counter
from pathlib import Path


class Solution:

    input_file: Path
    list_a: tuple[int, ...]
    list_b: tuple[int, ...]

    def __init__(self, input_file: str | Path):
        self.input_file = Path(input_file)
        self.list_a, self.list_b = self.process_input()

    def process_input(self) -> tuple[tuple[int, ...], tuple[int, ...]]:
        data = [[int(j) for j in i.split()] for i in self.input_file.open('r').readlines()]
        return zip(*data)

    def solve_part1(self) -> int:
        list_a, list_b = sorted(self.list_a), sorted(self.list_b)
        return sum(map(lambda a, b: abs(a - b), list_a, list_b))

    def solve_part2(self) -> int:
        list_a_count = Counter(self.list_a)
        list_b_count = Counter(self.list_b)
        return sum(list_a_count[i] * list_b_count[i] * i for i in list_a_count.keys())


if __name__ == "__main__":
    s = Solution("input.txt")
    print(s.solve_part1())
    print(s.solve_part2())

File: 2024/01/solution.nim
import std/sugar
import std/strutils
import std/sequtils
import std/algorithm
import std/sets
import system/iterators

proc readInput(file: string): (seq[int], seq[int]) =
  var k = collect:
    for line in lines file:
      var values: seq[string] = line.splitWhitespace
      (values[0].parseInt, values[1].parseInt)
  return k.unzip

let data = readInput("input.txt")

proc solvePart1(): int =
  var 
    listA: seq[int] = data[0]
    listB: seq[int] = data[1]
  
  listA.sort
  listB.sort
  
  for a in 0..listA.maxIndex:
    result += abs(listA[a] - listB[a])

proc solvePart2(): int =
  let unique: HashSet[int] = data[0].toHashSet * data[1].toHashSet
  for i in unique:
    result += data[0].count(i) * data[1].count(i) * i

echo solvePart1()
echo solvePart2()
